tincr/io/library/genCellLibrary.tcl

```diff
@@ -291,6 +291,110 @@ proc processMacroCell {c s fo} {
     }
 }
 
+proc doPorts { fo } {
+
+    puts "Doing ports..."
+
+    set lc [get_lib_cells]
+    set dict [uniqueSites]
+
+    puts $fo "    <cell>"
+    puts $fo "      <type>IPORT</type>"
+    puts $fo "          <is_port/>"
+    puts $fo "      <level>LEAF</level>"
+    puts $fo "      <pins>"
+    puts $fo "        <pin>"
+    puts $fo "          <name>PAD</name>"
+    puts $fo "          <direction>output</direction>"
+    puts $fo "        </pin>"
+    puts $fo "      </pins>"
+    puts $fo "      <bels>"
+    
+    dict for {type site} $dict {
+        set b [get_bels -of $site]
+        set ispad 0
+        foreach b [get_bels -of $site] {
+            if { [get_property TYPE $b] == "PAD" } {
+                if { [get_property NUM_OUTPUTS $site] > 0 } {
+                    puts $fo "        <bel>"
+                    puts $fo "          <id>"
+                    puts $fo "            <primitive_type>$type</primitive_type>"
+                    puts $fo "            <name>PAD</name>"
+                    puts $fo "          </id>"
+                    puts $fo "        </bel>"
+                }
+            }
+        }
+        
+    }
+    puts $fo "      </bels>"
+    puts $fo "    </cell>"
+
+    puts $fo "    <cell>"
```

Comment:
I would make a new function similar to the following:

```
proc print_port_header {type pin_direction} {
     puts $fo "      <type>$type</type>"
     puts $fo "      <is_port/>"
     puts $fo "      <level>LEAF</level>"
     puts $fo "      <pins>"
     puts $fo "        <pin>"
     puts $fo "          <name>PAD</name>"
     puts $fo "          <direction>$pin_direction</direction>"
     puts $fo "        </pin>"
     puts $fo "      </pins>"
     puts $fo "      <bels>"
}
```

Since the headers are so similar, we could just reuse the code.